File: Year-2023/LeetCode-February-Challenge/CapacityToShipPackagesWithinDDays.js
var shipWithinDays = function (weights, days) {

    const isValid = (weights, days, capacity) => {
        let requiredDays = 1;
        let currWeight = 0;

        for (let x of weights) {
            if (currWeight + x > capacity) {
                currWeight = 0;
                requiredDays++;
            }
            currWeight += x;
        }
        if (requiredDays > days) return false;
        else return true;
    }
    let low = 0, high = 0;
    for (let x of weights) {
        low = Math.max(low, x);
        high += x;
    }
    let res = high;
    while (low <= high) {
        let mid = low + Math.trunc((high - low) / 2);
        if (isValid(weights, days, mid)) {
            res = mid;
            high = mid - 1;
        } else low = mid + 1;
    }
    return res;
};
let weights = [3,2,2,4,1,4], days = 3;
console.log(shipWithinDays(weights, days));

File: Year-2023/LeetCode-February-Challenge/CapacityToShipPackagesWithinDDays.java
class Solution {
    private boolean isValid(int[] weights, int days, int capacity) {
        int requiredDays = 1;
        int currWeight = 0;
        for (int x : weights) {
            if (currWeight + x > capacity) {
                currWeight = 0;
                requiredDays++;
            }
            currWeight += x;
        }
        if (requiredDays > days)
            return false;
        else
            return true;
    }

    public int shipWithinDays(int[] weights, int days) {
        int low = 0, high = 0;
        for (int x : weights) {
            low = Math.max(low, x);
            high += x;
        }
        int res = high;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (isValid(weights, days, mid)) {
                res = mid;
                high = mid - 1;
            } else
                low = mid + 1;
        }
        return res;
    }
}

public class CapacityToShipPackagesWithinDDays {
    public static void main(String[] args) {
        int[] weights = { 3, 2, 2, 4, 1, 4 };
        int days = 3;
        Solution s = new Solution();
        System.out.println(s.shipWithinDays(weights, days));
    }
}